oak/server/rust/oak_runtime/src/channel.rs

```diff
@@ -0,0 +1,255 @@
+    });
+    let writer = ChannelWriter(ChannelRef::from_arc(chan.clone()));
+    let reader = ChannelReader(ChannelRef::from_arc(chan.clone()));
+    (writer, reader)
+}
+
+impl ChannelRef {
+    /// Internal method to take a thread-safe reference-counting pointer to a `Channel`.
+    fn from_arc(arc: Arc<Channel>) -> ChannelRef {
+        ChannelRef(arc)
+    }
+
+    /// Attempt to apply a function that takes a `&mut VecDeque<Message>` to the underlying message
+    /// store. Fails with `OakStatus::ERR_CHANNEL_CLOSED` if the underlying channel has been
+    /// orphaned.
+    fn try_write<F, T>(&self, f: F) -> Result<T, OakStatus>
+    where
+        F: FnOnce(&mut Messages) -> T,
+    {
+        if self.0.is_orphan() {
+            return Err(OakStatus::ERR_CHANNEL_CLOSED);
+        }
+
+        let mut messages = self.0.messages.write().unwrap();
+        Ok(f(&mut *messages))
+    }
+
+    /// Apply a function that takes a `&mut VecDeque<Message>` to the underlying message
+    /// store.
+    fn write<F, T>(&self, f: F) -> T
+    where
+        F: FnOnce(&mut Messages) -> T,
+    {
+        let mut messages = self.0.messages.write().unwrap();
+        f(&mut *messages)
+    }
+
+    /// Apply a function that takes a `&VecDeque<Message>` to the underlying message
+    /// store.
+    fn read<F, T>(&self, f: F) -> T
+    where
+        F: FnOnce(&Messages) -> T,
+    {
+        let messages = self.0.messages.read().unwrap();
+        f(&*messages)
+    }
+
+    /// Decrement the `Channel` writer counter
+    fn remove_writer(&self) {
+        self.0.writers.fetch_sub(1, SeqCst);
+    }
+
+    /// Decrement the `Channel` reader counter
+    fn remove_reader(&self) {
+        self.0.readers.fetch_sub(1, SeqCst);
+    }
+
+    /// Increment the `Channel` writer counter
+    fn add_writer(&self) -> usize {
+        self.0.writers.fetch_add(1, SeqCst)
+    }
+
+    /// Increment the `Channel` reader counter
+    fn add_reader(&self) -> usize {
+        self.0.readers.fetch_add(1, SeqCst)
+    }
+
+    /// Returns true if there are no messages in the channel _and_ if there is not at least one
+    /// writer and one reader.
+    fn is_empty_orphan(&self) -> bool {
+        self.0.is_orphan() && self.read(|messages| messages.len()) == 0
+    }
+}
+
+impl ChannelWriter {
+    /// Write a message to a channel. Fails with `OakStatus::ERR_CHANNEL_CLOSED` if the underlying
+    /// channel has been orphaned.
+    pub fn write(&self, msg: Message) -> Result<(), OakStatus> {
+        self.0.try_write(|messages| messages.push_back(msg))
+    }
+}
+
+impl Drop for ChannelWriter {
+    fn drop(&mut self) {
+        self.0.remove_writer();
+    }
+}
+
+impl Clone for ChannelWriter {
+    fn clone(&self) -> Self {
+        self.0.add_writer();
+        ChannelWriter(self.0.clone())
+    }
+}
+
+/// A helper type to determine if `try_read_message` was called with not enough `bytes_capacity`
+/// and/or `handles_capacity`.
+pub enum TakeMessage {
```

Comment:
How about `ReadResult` or `ReadStatus` or something like that?

Reply:
Changed to `ReadStatus`